atc/db/volume.go

```diff
@@ -353,37 +354,84 @@ func (volume *createdVolume) findWorkerBaseResourceTypeByBaseResourceTypeID(base
 }
 
 func (volume *createdVolume) InitializeResourceCache(resourceCache UsedResourceCache) error {
-	return volume.initializeResourceCache(resourceCache, volume.workerName)
-}
+	tx, err := volume.conn.Begin()
+	if err != nil {
+		return err
+	}
 
-func (volume *createdVolume) InitializeStreamedResourceCache(resourceCache UsedResourceCache, sourceWorkerName string) error {
-	return volume.initializeResourceCache(resourceCache, sourceWorkerName)
+	defer tx.Rollback()
+
+	usedWorkerBaseResourceType, found, err := WorkerBaseResourceType{
+		Name:       resourceCache.BaseResourceType().Name,
+		WorkerName: volume.workerName,
+	}.Find(tx)
+	if err != nil {
+		return err
+	}
+	if !found {
+		return ErrWorkerBaseResourceTypeDisappeared
+	}
+
+	initialized, err := volume.initializeResourceCache(tx, resourceCache, usedWorkerBaseResourceType.ID)
+	if err != nil {
+		return err
+	}
+	if !initialized {
+		// Another volume became the resource cache - don't commit the transaction
+		return nil
+	}
+
+	return tx.Commit()
 }
 
-// initializeResourceCache creates a worker resource cache and point current volume's
-// worker_resource_cache_id to the cache. When initializing a local generated resource
-// cache, then source worker is just the volume's worker.
-func (volume *createdVolume) initializeResourceCache(resourceCache UsedResourceCache, sourceWorkerName string) error {
+func (volume *createdVolume) InitializeStreamedResourceCache(resourceCache UsedResourceCache, sourceWorkerName string) error {
 	tx, err := volume.conn.Begin()
 	if err != nil {
 		return err
 	}
 
 	defer tx.Rollback()
 
+	sourceWorkerResourceCache, found, err := WorkerResourceCache{
+		WorkerName:    sourceWorkerName,
+		ResourceCache: resourceCache,
+	}.Find(tx)
+	if err != nil {
+		return err
+	}
+	if !found {
+		return ErrSourceResourceCacheDisappeared
```

Comment:
Is this a case that we want to error on? This could happen if while the volume was being streamed, the original worker (not necessarily the source worker) got a new base resource type version, invalidating the cache. Should we just give up and not initialize the resource cache, but not error?

Reply:
So that would mean that the volume was still able to stream correctly but we just can't initialize the resource cache for it? I think in that case we might be able to not error and just not initialize the cache

If we error that would mean that the `FindOrCreateContainer` for the container that we are trying to stream the volumes to will fail, and technically it shouldn't fail because it was able to stream the volume in correctly?

Reply:
Yeah - but it also means it's using a version of the volume that's likely "not the latest", since if the cache was invalidated, presumably there's a newer version of the base resource type. But it was the "latest" when the step started, so it's probably good enough. Will change to not error